### DISCUSSION TITLE
No interative window to run python script

### DISCUSSION THREAD

**Discussion by Naman-Agrawal19** [posit-dev/positron#4470] [Feedback]
This SS if from VS code. I can run python script line by line or whole script in an interactive window which is nothing but a jupyter notebook.

![Screenshot_20240823_231756](https://github.com/user-attachments/assets/fceb18a1-9cea-43ce-846e-0f75bf7c86ac)

<p> But i can't do this in Positron IDE. Kindly inbuilt this feature in IDE too,

**Comment by juliasilge** [comment]
Instead of the Jupyter interactive window, Positron has its own fully featured interactive console:

![console](https://github.com/user-attachments/assets/3e58d270-1658-43cc-850d-91d819a1bdd8)

You can send code line-by-line to the console using <kbd>Cmd/Ctrl</kbd>+<kbd>Enter</kbd>, or you can run the whole script or a selection:

![Screenshot 2024-08-23 at 4 40 22 PM](https://github.com/user-attachments/assets/33cfe876-5c49-4a3b-bf2c-4fe12a22c066)

Let me know if that doesn't answer your question!

**Reply by Naman-Agrawal19** [reply]
Ok, i got it, but interactive window support will be beneficial too i guess, atleast for me.

**Reply by juliasilge** [reply]
We'd be interested in hearing what features from the interactive window you miss when using the Positron console.

**Reply by Yayoarjona** [reply]
I think that to extend the use of the interactive window, the option of using the ipython interpreter (instead of python) could be added, this allows you to use the cells without having to use notebooks.

**Reply by juliasilge** [reply]
We do have support for code cells in Positron:

![code-cells](https://github.com/user-attachments/assets/30868cea-8195-4190-88cf-02d5ad7837bf)

You can use the keyboard shortcut or buttons to send code to the console.

**Reply by LucasFRChaves** [reply]
Hi @juliasilge,
I completely understand your point, especially since the console is fully connected to the "variables" and "plots" sections. However, I think what @PrimEGameR123 means is that the "interactive window" presents code outputs in a more sophisticated (and visually appealing) way compared to the console, making it easier and faster to identify the results of the code.
That said, the console wouldn’t be an obstacle to migrating from VSCode to Positron, but it would be nice to have that feature.

**Reply by juliasilge** [reply]
If you can share an example of code output in a Jupyter notebook that you are missing in the console, that would be super helpful!

**Reply by LucasFRChaves** [reply]
Hi @juliasilge ,

What I’m saying is that Jupyter notebooks in VS Code (VsJn) provide better visual output separation, making it easier to interpret results. As shown in Image 1 (Example 1), the output is clearly distinguished within each cell, allowing for a more intuitive understanding. Additionally, since each result is tied to a specific cell, it's easy to trace back the code that generated the output (Image 2).

Another advantage of VsJn is the ability to detach the interactive window from the main code window, which is incredibly useful when working with multiple screens, especially for debugging (Image 3).

If Positron has (or could implement) similar features, it would be extremely helpful!

![example_1](https://github.com/user-attachments/assets/6ec9de20-8395-4def-99bb-68496ca6c310)
![example_2](https://github.com/user-attachments/assets/752e140a-5fcd-49db-9223-5137f921f584)
![example_3](https://github.com/user-attachments/assets/3ca2f450-76fc-4140-a5d8-4baf2875216c)

**Reply by juliasilge** [reply]
We've got features such as this tracked in #243 and #3660, if you'd like to 👍 or subscribe to those.

I opened https://github.com/posit-dev/positron/issues/6225 to track and prioritize being able to open the console in a new window, such as for moving to a different monitor.

**Reply by ntluong95** [reply]
I think what they meant is not just about detaching the console, but also some treatment applied in the output. So far I have seen Positron has worked very hard to styling the input and output in the console.
One thing I would like to request again, is to improve dataframe printing, as requested in https://github.com/posit-dev/positron/issues/3802
At least it should look like rendered HTML table in Jupyter notebook in VSCode, or having some functions, for example a button to quick open the Data Explorer, exporting button for example.

**Reply by LukaNeg** [reply]
@juliasilge 
Not sure what stage this is at, so just re-pinging this thread. As a die-hard Rstudio lover that has had to convert over to VS-Code for  Python, I can definitely +1 the value of having output go to an interactive console as described above. Here are some of the advantages:
- The the code sent to the interactive console (IC) can be expanded or contracted (default contracted) if you need to see what you had sent to the console without it instantly taking up all the space. In the screenshots above, the three dots to the left of each block allow you to expand it.
- When running code that takes longer, there is a timer counting the minutes and seconds that it has been running for, and I can even go back to a previous run to remember "ahh that took 2 min to run, ok". You can see in the examples above the "0.0s" under each block.
- Any dataframes are displayed in a visually appealing format. I know we can explore the data in the data explorer, but sometimes I just want to run a df and see a quick summary of what the table looks like without having to open it up completely.
- The blocks of each code run are clearly defined "blocks" in a notebook-style
- You can delete previous blocks to organize what your output looks like. This isn't for creating a pretty notebook by any means, it's just for quick organization when, for example, you might want to compare some numbers that you ran earlier.
- In VSCode the plot outputs also get sent here, but in the case of Positron, you could leave that to go into the Plot viewer with maybe a link that you could click in the IC to take you directly to that plot in the viewer. Just a thought there, but there might be a better way to integrate IC with the plot viewer.
- Finally, the green check mark is a silly but really satisfying UX feature. It's very nice to see the code blocks with the check mark knowing nothing broke haha

Hope this is helpful! CC @LucasFRChaves

**Reply by LukaNeg** [reply]
I'll just add that the interactive console in VS Code comes from this: https://github.com/ipython/ipython

Maybe that helps for how you could integrate this into Positron

**Reply by juliasilge** [reply]
Thank you for this feedback, @LukaNeg! I added some of your thoughts to https://github.com/posit-dev/positron/issues/3660#issuecomment-3109047863 where we are tracking styling improvements to the Positron console.

I'll restate that we are all in on the fully interactive Positron console for use cases where you might in VS Code use the Jupyter interactive window. We really appreciate you highlighting what you miss from the Jupyter interactive window, so we can think about how to support those needs!

**Reply by LukaNeg** [reply]
Thank you so much! Really appreciate you taking this feedback into account 🙏

**Comment by FritzZuhl** [comment]
I very much miss the interactive window in VS Code. Please return it. I have been using the 'console' but it just is not the same.

**Reply by juliasilge** [reply]
Can you share more about what you miss from the Jupyter interactive window compared to the Positron console? Is it styling of inputs and output as described in https://github.com/posit-dev/positron/issues/3660? Or something else?

**Reply by jthomasmock** [reply]
Hi @FritzZuhl -- following up here, it would be great to hear about features you are missing or wish were available. 🙏
